### PR TITLE
mgr/cephadm: document setting custom monitoring templates for configuration files

### PR DESCRIPTION
**Pull request by p-se**
Fixes: https://tracker.ceph.com/issues/47651

Signed-off-by: Patrick Seidensal <pseidensal@suse.com>

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by bk201** [comment]
jenkins render docs

File: doc/cephadm/monitoring.rst [hunk 1]
```diff
@@ -140,6 +140,88 @@ For example
...
+- grafana_grafana.ini
+- prometheus_prometheus.yml
+
+You can look up the file templates that are currently used by cephadm in
+`src/pybind/mgr/cephadm/templates`:
+
+- services/alertmanager/alertmanager.yml.j2
+- services/grafana/ceph-dashboard.yml.j2
+- grafana/grafana.ini.j2
+- prometheus/prometheus.yml.j2
```
Comment (bk201) [inline doc/cephadm/monitoring.rst:182]:
```suggestion
- services/grafana/grafana.ini.j2
- services/prometheus/prometheus.yml.j2
```

**Review by bk201 — APPROVED** [review]

File: doc/cephadm/monitoring.rst [hunk 2]
```diff
@@ -140,6 +140,88 @@ For example
...
           ceph config rm mgr mgr/cephadm/container_image_prometheus
 
+Using custom configuration files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By overriding cephadm templates, it is possible to completely customize the
+configuration files for monitoring services.
+
+Internally, cephadm already uses `Jinja2
+<https://jinja.palletsprojects.com/en/2.11.x/>`_ template to generate the
```
Comment (LenzGr) [inline doc/cephadm/monitoring.rst:150]: Plural?
```suggestion
<https://jinja.palletsprojects.com/en/2.11.x/>`_ templates to generate the
```

File: doc/cephadm/monitoring.rst [hunk 3]
```diff
@@ -140,6 +140,88 @@ For example
 
           ceph config rm mgr mgr/cephadm/container_image_prometheus
 
+Using custom configuration files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By overriding cephadm templates, it is possible to completely customize the
+configuration files for monitoring services.
+
+Internally, cephadm already uses `Jinja2
```
Comment (LenzGr) [inline doc/cephadm/monitoring.rst:149]: Singular? "A template"?
```suggestion
Internally, cephadm already uses `Jinja2
```

File: doc/cephadm/monitoring.rst [hunk 4]
```diff
@@ -140,6 +140,88 @@ For example
...
+<https://jinja.palletsprojects.com/en/2.11.x/>`_ template to generate the
+configuration for all monitoring components. To be able to customize the
+configuration of Prometheus, Grafana or the Alertmanager it is possible to store
+a Jinja2 template that will be used for configuration generation instead. This
+template will be evaluated every time a service of that kind is deployed or
+reconfigured. That way the custom configuration is preserved and automatically
```
Comment (LenzGr) [inline doc/cephadm/monitoring.rst:155]:
```suggestion
reconfigured. That way, the custom configuration is preserved and automatically
```

File: doc/cephadm/monitoring.rst [hunk 5]
```diff
@@ -140,6 +140,88 @@ For example
...
+applied on deployments of future services.
```
Comment (LenzGr) [inline doc/cephadm/monitoring.rst:156]:
```suggestion
applied on future deployments of these services.
```

File: doc/cephadm/monitoring.rst [hunk 6]
```diff
@@ -140,6 +140,88 @@ For example
...
+  the custom template needs to be migrated *manually*.
+
+Option names
+""""""""""""
+
+The following templates for files that will be generated by cephadm can be
+overridden. These are the names to be used when storing with ``ceph config-key
+set``:
+
+- alertmanager_alertmanager.yml
```
Comment (LenzGr) [inline doc/cephadm/monitoring.rst:171]: I wonder if these should be written in monospaced font?

File: doc/cephadm/monitoring.rst [hunk 7]
```diff
@@ -140,6 +140,88 @@ For example
...
+- grafana_ceph-dashboard.yml
+- grafana_grafana.ini
+- prometheus_prometheus.yml
+
+You can look up the file templates that are currently used by cephadm in
+`src/pybind/mgr/cephadm/templates`:
```
Comment (LenzGr) [inline doc/cephadm/monitoring.rst:177]: I think single backticks for monospaced content are MarkDown, reStructuredText uses double backquotes.

File: doc/cephadm/monitoring.rst (hunk 3 above)
Comment (p-se) [reply doc/cephadm/monitoring.rst:149]: There's one template for each service. I'd argue for templates.

File: doc/cephadm/monitoring.rst [hunk 8]
```diff
@@ -140,6 +140,88 @@ For example
...
+.. code-block:: bash
+
+  # set the contents of ./prometheus.yml.j2 as template
+  ceph config-key set mgr/cephadm/services_prometheus_prometheus.yml \
+    -i $PWD/prometheus.yml.j2
+
+  # restart cephadm mgr module
+  ceph mgr module disable cephadm && \
+    sleep 2 && \
+    ceph mgr module enable cephadm
```
Comment (mgfritch) [inline doc/cephadm/monitoring.rst:220]:
```suggestion
  # restart cephadm mgr module
  ceph orch restart mgr
```
?

**Review by LenzGr — APPROVED** [review]
LGTM, thank you!

File: doc/cephadm/monitoring.rst (hunk 8 above)
Comment (p-se) [reply doc/cephadm/monitoring.rst]: I thought I did already :thinking: Maybe it was another PR. Thanks anyway for this great suggestion!

**Pull request merged**